### PR TITLE
Fix inscrutable einsum error in SpectralConv.forward when `separable=True`

### PR DESCRIPTION
**Pull request by dhpitt**
Changes to how the `SpectralConv` layer's tensor contraction function is chosen based on `implementation`, `factorization`, and `separable` to avoid einsum symbols and dense contract being incorrect.

Fixes issue #382 

### bug

Normally a `SpectralConv` is created with the option `separable=False` and its weights have shape `(in_channels, out_channels, modes_1, ...)`. When `separable=True`, the weights have shape `(channels, modes_1, ...)`. The indexing of weight into `weight[slices_w]` around lines 440-460 did not take this into account, leading to bugs in either the factorized tensor contraction einsums or in the slicing required before calling `self._contract`. 

### fix
In order to fix this I changed the indexing and slices mentioned above to ensure that slices are the proper length for the correct cases. I also made sure the proper tensor contraction function was being returned by `self.get_contract_fun` and added unit tests for all separable tensor cases.

### CODE REVIEW

File: neuralop/layers/spectral_convolution.py [hunk 1]
```diff
@@ -46,12 +47,6 @@ def _contract_dense(x, weight, separable=False):
         return tl.einsum(eq, x, weight)
 
 
-def _contract_dense_separable(x, weight, separable=True):
```
Comment (JeanKossaifi) [inline neuralop/layers/spectral_convolution.py:49]: Why was this removed?

File: neuralop/layers/spectral_convolution.py [hunk 2]
```diff
@@ -132,7 +127,7 @@ def _contract_tt(x, tt_weight, separable=False):
         return tl.einsum(eq, x, *tt_weight.factors)
 
 
-def get_contract_fun(weight, implementation="reconstructed", separable=False):
```
Comment (JeanKossaifi) [inline neuralop/layers/spectral_convolution.py:135]: Why did you remove the separable arg? How will we choose the right function?

**Review by JeanKossaifi — COMMENTED** [review]
Looks good overall though the separable function you removed should be a lot more efficient.

File: neuralop/layers/spectral_convolution.py (hunk 2 above)
Comment (dhpitt) [reply neuralop/layers/spectral_convolution.py:135]: All the factorized contraction functions have their own separable arg, which is passed at the call time for `self._contract`. In the reconstructed case, `_contract_dense_separable` is just elem-wise multiplication `x * w` as opposed to using einsum. I can add it back in
